### PR TITLE
Fix segment masks outside detection boxes

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary
- crop each generated instance mask to its detection box before rendering or returning per-instance mask data
- use Ultralytics-style raw logit thresholding for segmentation mask rendering
- speed up realtime segmentation by skipping per-instance nested mask arrays on camera frames
- unify realtime profiling so every task measures from frame acceptance through task postprocessing/result emission
- keep frame lifecycle ownership in the camera pipeline instead of the renderer
- share request-handler setup, timing, and Vision error handling across static and realtime inference paths
- remove the duplicate BasePredictor frame-buffer gate; VideoCapture owns frame admission
- add a regression test that verifies per-instance masks are zero outside the detection box

## Validation
- xcodebuild -project YOLOiOSApp/YOLOiOSApp.xcodeproj -scheme YOLOiOSApp -destination 'generic/platform=iOS Simulator' build
- swift test --filter PlotTests (fails before tests run: SwiftPM host build cannot import UIKit)
- xcodebuild ... -scheme YOLO test and -scheme YOLOiOSApp test (both schemes are not configured for the test action)

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🛠️ This PR refactors the iOS prediction pipeline to improve real-time stability, unify timing/error handling, and make segmentation masks more accurate.

### 📊 Key Changes
- 🔄 **Refactored shared prediction flow in `BasePredictor`**
  - Removed older per-frame buffer tracking from `BasePredictor`.
  - Added reusable helpers for creating Vision request handlers, running requests, and finishing timing.
  - Simplified `VNCoreMLRequest` setup so live-camera results are processed explicitly after inference.

- ⏱️ **Unified and improved inference timing**
  - Timing is now calculated consistently across classification, detection, pose, OBB, segmentation, and semantic segmentation.
  - Added `finishTiming()` and updated `updateTime()` so reported speed/FPS is more accurate and avoids stale values.

- 📷 **Safer real-time frame processing**
  - `VideoCapture` now checks `predictor.isUpdating` before starting a new frame.
  - In-progress state is set and cleared more reliably, reducing the chance of overlapping frame processing or UI-related race conditions.

- 🧹 **Cleaner failure handling**
  - Several predictors now explicitly reset update state when request results are missing or invalid.
  - Common request execution logic reduces duplicated `do/catch` blocks and standardizes logging.

- 🎭 **Improved segmentation mask behavior**
  - `generateCombinedMaskImage()` now crops each instance mask to its detection box before thresholding or returning mask data.
  - Default threshold changed from `0.5` to `0.0` in the mask generation helper, aligning better with the updated mask-processing flow.
  - Segmentation now supports skipping per-instance mask generation when only the combined mask is needed for live results.

- ✅ **CI and test updates**
  - GitHub Actions now tolerates missing coverage reports with `handle_no_reports_found: true`.
  - Added/updated tests for the new mask-cropping behavior and removed a test tied to the deleted buffer state.

### 🎯 Purpose & Impact
- 🚀 **Better real-time performance and stability**
  - Helps prevent duplicate or overlapping camera-frame inference, which should make the app smoother during live use.

- 📏 **More accurate speed reporting**
  - Users and developers should see inference timing that better reflects actual model execution across tasks.

- 🎯 **More precise segmentation outputs**
  - Instance masks are now constrained to their detected objects, reducing mask spill outside bounding boxes and improving visual quality.

- 🧩 **Easier maintenance**
  - Shared helper methods reduce repeated code, making the prediction pipeline easier to extend and less error-prone.

- 🛡️ **More resilient CI**
  - Builds are less likely to fail when coverage files are unavailable, improving workflow reliability for contributors.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/yolo-ios-app` 🚀 PR! This is an automated message to help with review readiness, and an engineer will assist you soon. Please review the checklist below:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/yolo-ios-app/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/yolo-ios-app` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Good fix overall: pre-cropping per-instance masks before composing addresses the bleed-outside-box behavior, and the regression test is useful. The main thing to correct before merging is the threshold change, which currently alters mask visibility semantics, plus the new `rounded()` box conversion which can drop narrow detections.

💬 Posted 2 inline comments

File: Sources/YOLO/Plot.swift [hunk 1]
```diff
@@ -196,7 +196,7 @@ func generateCombinedMaskImage(
   protos: MLMultiArray,  // shape: [1, C, H, W]
   inputWidth: Int,
   inputHeight: Int,
-  threshold: Float = 0.5,
+  threshold: Float = 0.0,
```
Comment (UltralyticsAssistant) [inline Sources/YOLO/Plot.swift:199]: ⚠️ **HIGH**: Changing the default mask threshold from `0.5` to `0.0` also changes the rendering behavior for all current `Segmenter` call sites. The mask values copied into `probabilityMasks` here are probabilities, so `> 0.0` will effectively light up almost every in-box pixel instead of preserving the previous segmentation cutoff. Cropping can be applied before filtering without lowering the visible-mask threshold.

**Suggested change:**
```suggestion
  threshold: Float = 0.5,
```

File: Sources/YOLO/Plot.swift [hunk 2]
```diff
@@ -272,21 +272,43 @@ func generateCombinedMaskImage(
   let outputHeight = Int(outputRect.height)
   guard outputWidth > 0, outputHeight > 0 else { return nil }
 
+  // Match Ultralytics process_mask(): crop each mask to its detection box before thresholding
+  // or returning per-instance data.
+  var maskBoxes: [(x1: Int, y1: Int, x2: Int, y2: Int)] = []
+  maskBoxes.reserveCapacity(N)
+  for (box, _, _, _) in detectedObjects {
+    let x1 = Int((Float(box.minX) * scaleX).rounded())
```
Comment (UltralyticsAssistant) [inline Sources/YOLO/Plot.swift:280]: 💡 **MEDIUM**: Using `rounded()` for both box minima and maxima can collapse small or fractional boxes to zero area after scaling. For example, a box spanning `0.6...1.4` rounds to `1...1` and is skipped by the `x1 < x2` guard, whereas it should still cover one pixel. Min edges should be floored and max edges ceiled (or the previous truncation behavior kept) to avoid dropping narrow masks.

**Comment by UltralyticsAssistant** [comment]
Merged! 🎉 Thanks @glenn-jocher for this thoughtful refactor and the clear improvements to the iOS prediction pipeline.

As Leonardo da Vinci said, *“Simplicity is the ultimate sophistication.”* This PR is a great example of that in practice: cleaner shared prediction flow, more reliable real-time frame handling, more accurate timing, and sharper segmentation behavior all add up to a smoother, more maintainable app.

Really appreciate the care put into stability, mask quality, and CI resilience here—this kind of work makes a big difference for both contributors and end users.
